src/main/scala/com/fulcrumgenomics/bam/ConsensusCaller.scala

```diff
@@ -0,0 +1,369 @@
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package com.fulcrumgenomics.bam
+
+import com.fulcrumgenomics.bam.ConsensusCallerOptions._
+import com.fulcrumgenomics.util.PhredValue.{OneProbability, ZeroPhred, ZeroProbability}
+import com.fulcrumgenomics.util.{PhredValue, ProgressLogger}
+import dagr.commons.CommonsDef._
+import htsjdk.samtools._
+import htsjdk.samtools.util.SequenceUtil
+
+import scala.collection.mutable
+import scala.collection.mutable.ListBuffer
+
+object ConsensusCallerOptions {
+  /** Various default values for the consensus caller. */
+  val DefaultAttribute: String                       = "RX"
+  val DefaultErrorRatePreUmi: PhredValue             = 50
+  val DefaultErrorRatePostUmi: PhredValue            = 0
+  val DefaultMaxBaseQuality: PhredValue              = 40
+  val DefaultBaseQualityShift: PhredValue            = 10
+  val DefaultMinConsensusBaseQuality: PhredValue     = 13
+  val DefaultMinReads: Int                           = 1
+  val DefaultMinMeanConsensusBaseQuality: PhredValue = 13
+}
+
+/** Holds the parameters/options for consensus calling. */
+case class ConsensusCallerOptions(attribute: String                       = DefaultAttribute,
+                                  errorRatePreUmi: PhredValue             = DefaultErrorRatePreUmi,
+                                  errorRatePostUmi: PhredValue            = DefaultErrorRatePostUmi,
+                                  maxBaseQuality: PhredValue              = DefaultMaxBaseQuality,
+                                  baseQualityShift: PhredValue            = DefaultBaseQualityShift,
+                                  minConsensusBaseQuality: PhredValue     = DefaultMinConsensusBaseQuality,
+                                  minReads: Int                           = DefaultMinReads,
+                                  minMeanConsensusBaseQuality: PhredValue = DefaultMinMeanConsensusBaseQuality
+                                 )
+
+object ConsensusCaller {
+  /** Creates a consensus read from the given read and quality string tuples.  If no consensus read was created, None is returned. */
+  def fromBasesAndQualities(basesAndQualities: Seq[(String, String)],
+                            options: ConsensusCallerOptions = new ConsensusCallerOptions()
+                           ): Option[(String, String)] = {
```

Comment:
Maybe we should use a case class for `ConsensusRead` instead of a tuple?  Works the same, but will give us place to hang extra stats or anything else we want.